src/schema.ts

```diff
@@ -0,0 +1,385 @@
+
+/** Represents the dropping of a single marble with an associated tick. */
+export type TickedDropEvent = CoreDropEvent & DropEvent;
+
+/** Represents information common to all drop events. */
+export interface CoreDropEvent {
+	/** The tick (pulse) that the marble is to be dropped on. */
+	tick: number;
+}
+/** Represents the dropping of a single bass marble. */
+export interface BassDropEvent {
+	kind: "bass";
+	/** The string (not note) to drop the marble onto. */
+	string: BassString;
+	/**
+	 * The fret to be played when this marble hits.
+	 *
+	 * Even though this technically requires user input, we opted to specify the fret here
+	 * because it's much simpler than requiring a separate fret pressed event in the program.
+	 */
+	fret: number;
+}
+/** Represents the dropping of a single drum marble. */
+export interface DrumDropEvent {
+	kind: "drum";
+	/** The drum to drop the marble onto. */
+	drum: DrumType;
+}
+/** Represents the dropping of a single vibraphone marble. */
+export interface VibraphoneDropEvent {
+	kind: "vibraphone";
+	/** The channel (key) to drop the marble onto.
+	 *
+	 * To figure out which note it will play, the [[VibraphoneState.notes]]
+	 * property must be referenced.
+	 * ```typescript
+	 * const vDropEvent: VibraphoneDropEvent;
+	 * const vState: VibraphoneState;
+	 * const noteHit = vState.notes[vDropEvent.channel];
+	 * ```
+	 */
+	channel: VibraphoneChannel;
+}
+
+/**
+ * Represents a single performance of the associated [[Program]].
+ */
+export interface Performance {
+	/** Information about this performance. */
+	metadata: PerformanceMetadata;
+	/** The program associated with this performance. */
+	program: Program;
+	/**
+	 * The state of the machine at the start of the performance.
+	 *
+	 * Note that this should not be changed during the performance.
+	 * Use [[Program.state]] if you want to keep track of the current
+	 * state of the machine. This means that [[Performance.initialState]]
+	 * should be copied into [[Program.state]] at tick 0 of Performance playback.
+	 */
+	readonly initialState: State;
+	/**
+	 * The events that make up this performance. As with [[Program.dropEvents]],
+	 * these events must be in ascending order.
+	 */
+	events: TimedEvent[];
+}
+/** Metadata for performance */
+export interface PerformanceMetadata {
+	/**
+	 * The title of the performance.
+	 * This doesn't have to be the same as the title of the program.
+	 */
+	title: string;
+	/**
+	 * The author/performer of the performance.
+	 * This doesn't have to be the same as the author of the program.
+	 */
+	author: string;
+}
+
+/**
+ * An untimed event that changes the state of one channel
+ * or manually drops a marble.
+ *
+ * Events only change one thing. Multiple simultaneous changes must
+ * be represented by multiple events with the same tick.
+ */
+export type Event =
+	| DropEvent
+	| MachineEvent
+	| VibraphoneEvent
+	| HihatMachineEvent
+	| HihatEvent
+	| BassEvent;
```

Comment:
Some extra comments maybe?
		| DropEvent // Performances only store manual drops
		| MachineEvent // mutes
		| VibraphoneEvent // vibrato (& dampener when implemented)
		| HihatMachineEvent // setting
		| HihatEvent // open or closed
		| BassEvent; // capo change

Although this may make things harder to update if things change...

Reply:
If people want to know what properties the different events have, they can simply look it up in the specific Event definition.
`Performances only store manual drops` might be good to add, but in the Program's doc string.

Reply:
Update: There is actually an issue with this approach; see PR #14.